BFGS/bfgs.cpp

```diff
@@ -0,0 +1,102 @@
+// Test BFGS method on problems (serial)
+#include <iostream>
+#include <cmath>
+#include <fstream>
+#include "../test/testfun.hpp"
+#include <Eigen/Dense>
+
+using namespace std;
+
+template<typename ObjFunc> // Use the parameter 'max_bound' to define edge of hypercube to optimse in
+Eigen::VectorXd bfgs(ObjFunc f, const int max_iter, const Eigen::VectorXd& int_val,
+  const double rel_sol_change_tol, const double grad_nrm_tol, const double gam,
+   const double max_bound){
+    // Initialise vectors to hold xk, xk+1 and print initial value.
+    Eigen::VectorXd step_prev = int_val, step;
+
+    // Initialise dimension and gradient vectors of xk, xk+1 respectively
+    const int dim = int_val.size();
+    Eigen::VectorXd df_prev(dim), df(dim), s(dim), q(dim), w(dim);
+
+    // Useful constants for updates
+    double st_q, qt_d_q;
+
+    // Initialise relative changes to large number
+    double rel_sol_change = 2*rel_sol_change_tol;
+    double grad_nrm = 2*grad_nrm_tol;
+
+    // Initialise matrix D as the identity matrix
+    Eigen::MatrixXd d = Eigen::MatrixXd::Identity(dim, dim);
```

Comment:
Minor but keep matrices as capital letters just to make them stand out a bit more, I'll change this now